### DISCUSSION TITLE
docs gives vague error if json schema not defined on custom type with pydantic v2 + misc bugs

### DISCUSSION THREAD

**Discussion by GuillaumeQuenneville** [fastapi/fastapi#9865] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated, Any, Callable
from fastapi import FastAPI
import numpy as np
from pydantic import BaseModel
from pydantic_core import core_schema
import numpy.typing as npt

app = FastAPI()


class _NDArrayAnnotation:
    @classmethod
    def __get_pydantic_core_schema__(
        cls,
        _source_type: Any,
        _handler: Callable[[Any], core_schema.CoreSchema],
    ) -> core_schema.CoreSchema:
        return core_schema.is_instance_schema(np.ndarray)


NDArray = Annotated[npt.NDArray[Any], _NDArrayAnnotation]


class Mytype(BaseModel):
    array: NDArray


@app.get("/")
def read_root(body: Mytype):
    return {"Hello": "World"}
```

### Description

Hi all,

I'm not 100% confident of my ability to triage whether these are pydantic issues or fastapi issues so apologies in advance.

I've been playing with v2 and can't seem to figure what to implement for the openApi schema to generate properly. I've attached a minimum examples.

I have an numpy array type with a special format in json.

When viewing the docs I get the following error

<img width="1423" alt="image" src="https://github.com/tiangolo/fastapi/assets/32133082/7751da07-fcce-4aaf-b34e-dfe9f94ceec8">

which is vague and is caused by not implementing the `__get_pydantic_json_schema__` in the custom type.

Trying to print the `print(MyType.model_json_schema())` for the type gives
`    raise PydanticInvalidForJsonSchema(f'Cannot generate a JsonSchema for {error_info}')` 
So maybe fastapi should catch the error and display something useful?

I've written two other possible bugs in the additional context sections as well as an example with `__get_pydantic_json_schema__` implemented.

### Operating System

Linux

### FastAPI Version

0.100.0

### Python Version

Python 3.11.1

### Additional Context

one last issue.

core_schema.int_schema seems to resolve to `<any>` instead of number in the docs

<img width="726" alt="image" src="https://github.com/tiangolo/fastapi/assets/32133082/616a7561-b5ae-43bf-88f2-9bed5aed893c">

core_schema.float_schema resolves to `number` as expected

The generate schema.json have
```json
{
items: {
type: "integer"
},
type: "array"
}
```

integer in the type as expected

**Comment by GuillaumeQuenneville** [comment]
I saw an issue with the Input__1 in the other pydantic v2 thread but it isn't the same as this one.

Also they said that their issue was fixed in 
pydantic                  2.0.2
pydantic_core             2.1.2

but those are the versions that I am working on.

**Comment by samuelcolvin** [comment]
If you think there's an issue on the pydantic side, please create a bug issue on pydantic with and MRE and we'll look into it.

The issue with `core_schema.int_schema` is probably related to https://github.com/pydantic/pydantic/issues/6647
